lib/linguist/strategy/modeline.rb

```diff
@@ -0,0 +1,31 @@
+module Linguist
+  module Strategy
+    class Modeline
+      # Public: Detects language based on Vim and Emacs modelines
+      #
+      # blob               - An object that quacks like a blob.
+      #
+      # Examples
+      #
+      #   Modeline.call(FileBlob.new("path/to/file"))
+      #
+      # Returns an Array with one Language if the blob has a shebang with a valid
+      # interpreter, or empty if there is no shebang.
+      def self.call(blob, _ = nil)
+        if language = Language.find_by_alias(modeline(blob.data))
+          return [language]
+        else
+          return []
+        end
+      end
+
+      # Public: Get the modeline from the first n-lines of the file
+      #
+      # Returns a String or nil
+      def self.modeline(data)
+        data.lines.first(5).any? { |l| l.match(/\W(?:filetype=|ft=|-\*- mode:|-\*-)\s*(\w+)/) }
```

Comment:
Ready for more insanity? This enforces that the vim syntax is closed with vim syntax, and emacs is closed with emacs syntax. It also uses the extended regular expression syntax (`/x`), which ignores whitespace in the regex so you can split it across multiple lines.

```ruby
/
  (?:
    (-\*- \s* (?:mode:)? \s*) |                  # $1: Emacs
    (\/\* \s* vim: \s* set \s* (?:ft|filetype)=) # $2: Vim
  )
  (\w+)                                          # $3: language
  (?:
    (?(1)                                        # If $1 matched...
      ;?\s* -\*- |                               # then close Emacs syntax
      : \s* \*\/                                 # otherwise close Vim syntax
    )
  )
/x
```